File: android/src/main/java/com/taobao/power_image/request/PowerImageRequestManager.java
package com.taobao.power_image.request;

import static com.taobao.power_image.request.PowerImageBaseRequest.RENDER_TYPE_EXTERNAL;
import static com.taobao.power_image.request.PowerImageBaseRequest.RENDER_TYPE_TEXTURE;

import java.lang.ref.WeakReference;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import io.flutter.view.TextureRegistry;

/**
 * created by Muke on 2021/7/21
 */
public class PowerImageRequestManager {

    private Map<String, PowerImageBaseRequest> requests;
    private WeakReference<TextureRegistry> textureRegistryWrf;

    private PowerImageRequestManager() {
        requests = new HashMap<>();
    }

    private static class Holder {
        private final static PowerImageRequestManager instance = new PowerImageRequestManager();
    }

    public static PowerImageRequestManager getInstance() {
        return Holder.instance;
    }

    public void configWithTextureRegistry(TextureRegistry textureRegistry) {
        textureRegistryWrf = new WeakReference<>(textureRegistry);
    }

    public List<Map<String, Object>> configRequestsWithArguments(List<Map<String, Object>> list) {
        List<Map<String, Object>> results = new ArrayList<>();
        if (list == null || list.isEmpty()) {
            return results;
        }
        for (int i = 0; i < list.size(); i++) {
            Map<String, Object> arguments = list.get(i);
            String renderType = (String) arguments.get("renderingType");
            PowerImageBaseRequest request;
            if (RENDER_TYPE_EXTERNAL.equals(renderType)) {
                request = new PowerImageExternalRequest(arguments);
            } else if (RENDER_TYPE_TEXTURE.equals(renderType)) {
                request = new PowerImageTextureRequest(arguments, textureRegistryWrf.get());
            } else {
                continue;
            }
            requests.put(request.requestId, request);
            boolean success = request.configTask();
            Map<String, Object> requestInfo = request.encode();
            requestInfo.put("success", success);
            results.add(requestInfo);
        }
        return results;
    }

    public void startLoadingWithArguments(List arguments) {
        if (arguments == null || arguments.isEmpty()) {
            return;
        }
        for (int i = 0; i < arguments.size(); i++) {
            Map arg = (Map) arguments.get(i);
            String requestId = (String) arg.get("uniqueKey");
            PowerImageBaseRequest request = requests.get(requestId);
            request.startLoading();
        }
    }

    public List<Map<String, Object>> releaseRequestsWithArguments(List arguments) {
        List<Map<String, Object>> results = new ArrayList<>();
        if (arguments == null || arguments.isEmpty()) {
            return results;
        }
        for (int i = 0; i < arguments.size(); i++) {
            Map arg = (Map) arguments.get(i);
            String requestId = (String) arg.get("uniqueKey");
            PowerImageBaseRequest request = requests.get(requestId);
            if (request != null) {
                requests.remove(requestId);
                boolean success = request.stopTask();
                Map<String, Object> requestInfo = request.encode();
                requestInfo.put("success", success);
                results.add(requestInfo);
            }
        }
        return results;
    }
}

File: ios/Classes/request/PowerImageRequestManager.m
//
//  PowerImageRequestManager.m
//  power_image
//
//  Created by 王振辉 on 2021/7/14.
//

#import "PowerImageRequestManager.h"
#import <Flutter/Flutter.h>
#import "PowerImageBaseRequest.h"
#import "PowerImageTextureRequest.h"
#import "PowerImageExternalRequest.h"

@interface PowerImageRequestManager ()

@property (nonatomic, strong) NSMutableDictionary <NSString *, PowerImageBaseRequest *> * requests;
@property (nonatomic, weak) id<FlutterTextureRegistry> textureRegistry;

@end

@implementation PowerImageRequestManager


+ (instancetype)sharedInstance {
    static PowerImageRequestManager *instance;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        instance = [[PowerImageRequestManager alloc] init];
    });
    return instance;
}

- (instancetype)init
{
    self = [super init];
    if (self) {
        _requests = [[NSMutableDictionary alloc] init];
    }
    return self;
}

- (void)configWithTextureRegistry:(id<FlutterTextureRegistry>)textureRegistry {
    self.textureRegistry = textureRegistry;
}

/**
 * 初始化任务
 */
- (NSArray *)configRequestsWithArguments:(NSArray *)list{
    NSMutableArray *results = [[NSMutableArray alloc] init];
    for (NSInteger i = 0; i < list.count; i++) {
        NSDictionary *arguments = list[i];

        NSString *renderType = [self _renderingType:arguments];
        PowerImageBaseRequest *request;
        if ([renderType isEqualToString:PowerImageRequestRenderTypeExternal]) {
            request = [[PowerImageExternalRequest alloc] initWithArguments:arguments];
        }else if ([renderType isEqualToString:PowerImageRequestRenderTypeTexture]) {
            request = [[PowerImageTextureRequest alloc] initWithArguments:arguments textureRegistry:self.textureRegistry];
        }else {
            continue;
        }
        
        self.requests[request.uniqueKey] = request;
        BOOL success = [request configTask];
        
        NSMutableDictionary *requestInfo = [request encode];
        requestInfo[@"success"] = @(success);
        [results addObject:requestInfo];
    }
    return results;
}

/**
* 取消任务
*/
- (NSArray *)releaseRequestsWithArguments:(NSArray *)list{
    NSMutableArray *results = [[NSMutableArray alloc] init];
    for (NSInteger i = 0; i < list.count;  i++) {
        NSDictionary *arguments = list[i];
        NSString *uniqueKey = [self _uniqueKey:arguments];
        PowerImageBaseRequest *request = self.requests[uniqueKey];
        [self.requests removeObjectForKey:uniqueKey];
        if (!request) {
            continue;
        }
        BOOL success = [request stopTask];
        NSMutableDictionary *requestInfo = [request encode];
        requestInfo[@"success"] = @(success);
        [results addObject:requestInfo];
    }
    return results;
}

 /**
  * 开始图片加载的逻辑
  * 加载结果异步上报
  */
- (void)startLoadingWithArguments:(NSArray *)list {
    if (list.count == 0) {
        return;
    }
    
    for (NSInteger i = 0; i < list.count; i++) {
        NSDictionary *arguments = list[i];
        NSString *uniqueKey = [self _uniqueKey:arguments];
        PowerImageBaseRequest *request = self.requests[uniqueKey];
        [request startLoading];
    }
}

- (NSString *)_uniqueKey:(NSDictionary *)arguments {
    return arguments[@"uniqueKey"];
}

- (NSString *)_renderingType:(NSDictionary *)arguments {
    return arguments[@"renderingType"];
}

@end
